system/security/shared/caching.cpp

```diff
@@ -672,20 +680,42 @@ bool CPermissionsCache::queryPermsManagedFileScope(ISecUser& sec_user, const cha
 
 SecAccessFlags CPermissionsCache::queryDefaultPermission(ISecUser& user)
 {
-    const char *username = user.getName();
 
-    SecAccessFlags defaultPermission = SecAccess_None;
-    auto it = m_userDefaultFileScopePermissions.find(username);
-    if (it == m_userDefaultFileScopePermissions.end())
+    if (m_useLegacyDefaultFileScopePermissionCache)
     {
-        defaultPermission = m_secMgr->queryDefaultPermission(user);
-        std::string userName(username);
-        m_userDefaultFileScopePermissions.insert(std::pair<std::string, SecAccessFlags>(userName, defaultPermission));
-        DBGLOG("Added user '%s' to default file scope permissions with access %s(%d)", username, getSecAccessFlagName(defaultPermission), defaultPermission);
+        if (m_defaultPermission == SecAccess_Unknown)
+        {
+            if (m_secMgr)
+                m_defaultPermission = m_secMgr->queryDefaultPermission(user);
+            else
+                m_defaultPermission = SecAccess_None;
+
+            DBGLOG("Legacy default file scope permission set to %s(%d) for all users, based on User '%s'", getSecAccessFlagName(m_defaultPermission),
+                   m_defaultPermission, user.getName());
+        }
+        return m_defaultPermission;
     }
-    else
+
+    const char *username = user.getName();
+    SecAccessFlags defaultPermission = SecAccess_None;
+    if (m_secMgr)
     {
-        defaultPermission = it->second;
+        ReadLockBlock defaultScopesReadLock(m_defaultScopesRWLock);
+        auto it = m_userDefaultFileScopePermissions.find(username);
+        if (it == m_userDefaultFileScopePermissions.end())
+        {
+            defaultScopesReadLock.clear();
+            WriteLockBlock defaultScopesWriteLock(m_defaultScopesRWLock);
```

Comment:
I'm a bit surprised there isn't a lock.changeMode() that supports converting a read lock to a write lock, without releasing the read lock 1st.

Reply:
minor: After gaining this exclusive lock, it's probably worth re-checking m_userDefaultFileScopePermissions.find(username) , since another thread may have got in, and looked it up, and added it.